src/Generating/PieceModifier.cpp

```diff
@@ -0,0 +1,410 @@
+		} \
+	} while (false)
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+/** A modifier which is pseudo-randomly replacing blocks to other types and metas. */
+class cPieceModifierRandomizeBlocks:
+	public cPiece::cPieceModifier
+{
+public:
+	cPieceModifierRandomizeBlocks(void) :
+		m_Seed()
+	{
+	}
+
+	virtual bool InitializeFromString(const AString & a_Params, bool a_LogWarnings) override
+	{
+		AString Params = a_Params;
+
+
+		/** BlocksToReplace parsing */
+		auto idxPipe = Params.find('|');
+		if (idxPipe != AString::npos)
+		{
+			AString blocksToReplaceStr = Params.substr(0, idxPipe);
+			auto blocksToReplace = StringSplitAndTrim(blocksToReplaceStr, ",");
+			for (size_t i = 0; i < blocksToReplace.size(); i++)
+			{
+				BLOCKTYPE blockType = static_cast<BLOCKTYPE>(BlockStringToType(blocksToReplace[i]));
+				if ((blockType == E_BLOCK_AIR) && !NoCaseCompare(blocksToReplace[i], "Air"))
+				{
+					CONDWARNING(a_LogWarnings, "Cannot parse block type from string \"%s\"!", blocksToReplace[i].c_str());
+					return false;
+				}
+				m_BlocksToReplace[blockType] = 1;
+			}
+
+			if (m_BlocksToReplace.size() == 0)
+			{
+				CONDWARNING(a_LogWarnings, "You must specify at least one block to replace%s!", "");
+				return false;
+			}
+
+			Params = Params.substr(idxPipe + 1, Params.length() - 1);
+		}
+
+
+		/** Meta params parsing */
+		auto idxSqBracketStart = Params.find('[');
+		auto idxSqBracketStartLast = Params.find_last_of('[');
+
+		bool isMultiMeta = false;
+		if ((idxSqBracketStart != idxSqBracketStartLast) && (idxSqBracketStartLast != Params.length()-1))
+		{
+			// Meta per block type
+			isMultiMeta = true;
+		}
+		else
+		{
+			auto idxSqBracketEnd = Params.find(']');
+			if ((idxSqBracketEnd == Params.length()-1) && (idxSqBracketStart != AString::npos))
+			{
+				AString metaParamsStr = Params.substr(idxSqBracketStart + 1, Params.length()-idxSqBracketStart-2);
+				std::array<int, 4> metaParamsInt;
+				if (!ParseMeta(metaParamsStr, metaParamsInt, a_LogWarnings))
+				{
+					return false;
+				}
+
+				m_MinMeta = metaParamsInt[0];
+				m_MaxMeta = metaParamsInt[1];
+				m_MinNoiseMeta = metaParamsInt[2];
+				m_MaxNoiseMeta = metaParamsInt[3];
+
+				Params = Params.substr(0, idxSqBracketStart);
+			}
+		}
+
+
+		// BlocksToRandomize parsing
+		auto BlocksToRandomize = StringSplitAndTrim(Params, ";");
+		for (size_t i = 0; i < BlocksToRandomize.size(); i++)
+		{
+			AString block = BlocksToRandomize[i];
+
+			cBlockWithWeight Block{};
+
+			if (isMultiMeta)
+			{
+				auto sqBrStart = block.find('[');
+				if (sqBrStart != AString::npos)
+				{
+					auto sqBrEnd = block.find(']');
+					if (sqBrEnd != block.size() - 1)
+					{
+						CONDWARNING(a_LogWarnings, "If present, block meta params must be at the end of block to randomize definition \"%s\"!", block.c_str());
+						return false;
+
+					}
+					AString metaParamsStr = block.substr(sqBrStart + 1, block.size()-sqBrStart-2);
```

Comment:
Could you add spaces between the minus operators? I'm not sure why the style checker doesn't see it.